### PR TITLE
Features/black

### PR DESCRIPTION
**Pull request by liuhaox1**

### CODE REVIEW

File: core/peers_meter.go [hunk 1]
```diff
@@ -24,36 +24,67 @@ import (
...
+	}
 }
 func (m *peerMeter) decreaseTimeout() {
 	m.mu.Lock()
 	defer m.mu.Unlock()
-	m.timeoutMeter--
-	if m.timeoutMeter < 0 {
-		m.timeoutMeter = 0
-	}
+	m.timeoutMeter = 0
```
Comment (bird512) [inline core/peers_meter.go:87]: function name deceaseTime mismatch what it does inside (m.timeoutMeter = 0)

File: core/peers_meter.go [hunk 2]
```diff
@@ -24,36 +24,67 @@ import (
 )
```
Comment (darren0718) [inline core/peers_meter.go:25]: peerMeter needs unit test

File: core/transaction_pool.go [hunk 3]
```diff
@@ -107,15 +118,21 @@ func (pool *txPool) AddTransaction(tx *types.Transaction) (bool, error) {
...
+func (pool *txPool) AddTransactions(txs []*types.Transaction, from txSource) (evilCount int){
 	if nil == txs || 0 == len(txs) {
 		return
 	}
+
 	for _, tx := range txs {
 		// this error can be ignored
-		_, _ = pool.tryAddTransaction(tx, from)
+		_, err := pool.tryAddTransaction(tx, from)
+		if _,ok:=evilErrorMap[err];ok{
```
Comment (darren0718) [inline core/transaction_pool.go]: Is error addressable? what about err == nil ?

File: core/tx_sync.go [hunk 4]
```diff
@@ -216,6 +254,10 @@ func (ts *txSyncer) getOrAddCandidateKeys(id string) *peerTxsHashs {
 
 func (ts *txSyncer) onTxNotify(msg notify.Message) {
 	nm := notify.AsDefault(msg)
+	if peerManagerImpl.getOrAddPeer(nm.Source()).isEvil(){
+		ts.logger.Warnf("tx sync this source is is in evil...source is is %v\n",nm.Source())
+		return
+	}
```
Comment (darren0718) [inline core/tx_sync.go:236]: why count > txMaxReceiveLimit ? should be txMaxNotifyPerTime

File: core/tx_sync.go [hunk 5]
```diff
@@ -345,12 +414,41 @@ func (ts *txSyncer) onTxReq(msg notify.Message) {
...
+		ts.syncTxComplete(nm.Source(), false)
+	}()
+
 	txs, e := types.UnMarshalTransactions(nm.Body())
 	if e != nil {
 		ts.logger.Errorf("Unmarshal got transactions error:%s", e.Error())
 		return
 	}
 
+	if len(txs) > txReceiveMaxLimit{
```
Comment (darren0718) [inline core/tx_sync.go]: should be txMaxNotifyPerTime.
txReceiveMaxLimit(1000) too much!!!

**Review by darren0718 — CHANGES_REQUESTED** [review]

**Review by darren0718 — DISMISSED** [review]

File: core/block_sync.go [hunk 6]
```diff
@@ -119,25 +119,37 @@ func (bs *blockSyncer) getBestCandidate(candidateID string) (string, *topBlockIn
...
 	}
-	maxTop := bs.candidatePool[candidateID]
+	if temp == ""{
+		return "", nil
+	}
+	maxTop := bs.candidatePool[temp]
 	if maxTop == nil {
 		return "", nil
 	}
+	if candidateID == ""{
```
Comment (sesamegu) [inline core/block_sync.go]: why should candidateID must be null?

File: core/peers_meter.go [hunk 7]
```diff
@@ -24,37 +24,68 @@ import (
...
 
 const (
-	evilTimeoutMeterMax = 3
+	evilTimeoutMeterMax = 5
+	evilTimAdd = 10
 	maxReqBlockCount    = 16
+	maxEvilCount    = 100
+	maxEvilExpireTime = "1m"
```
Comment (sesamegu) [inline core/peers_meter.go]: maxEvilExpireTime is not a good name

File: core/transaction_pool.go [hunk 8]
```diff
@@ -38,13 +39,23 @@ const (
 	// Maximum size per transaction
 )
 
-// gasLimitMax expresses the max gasLimit of a transaction
-var gasLimitMax = new(types.BigInt).SetUint64(500000)
+var evilErrorMap map[error]struct{} = map[error]struct{}{
```
Comment (sesamegu) [inline core/transaction_pool.go]: type can be omitted

File: core/tx_sync.go [hunk 9]
```diff
@@ -79,6 +84,37 @@ func (ptk *peerTxsHashs) reset() {
 	ptk.txHashs = common.MustNewLRUCache(txPeerMaxLimit)
 }
 
+func (ptk *peerTxsHashs) resetSendHashs() {
+	ptk.sendHashs = common.MustNewLRUCache(txPeerMaxLimit)
+}
+
+func (ptk *peerTxsHashs) checkReceivedHashsInHitRate(txs []*types.Transaction)bool {
```
Comment (sesamegu) [inline core/tx_sync.go]: add unit test

**Review by sesamegu — APPROVED** [review]
ok

**Pull request merged**